noir-projects/noir-contracts/contracts/nft_contract/src/main.nr

```diff
@@ -0,0 +1,191 @@
+    }
+
+    #[aztec(public)]
+    #[aztec(view)]
+    fn public_get_symbol() -> pub FieldCompressedString {
+        storage.symbol.read_public()
+    }
+
+    #[aztec(private)]
+    #[aztec(view)]
+    fn private_get_symbol() -> pub FieldCompressedString {
+        storage.symbol.read_private()
+    }
+
+    #[aztec(public)]
+    #[aztec(view)]
+    fn admin() -> Field {
+        storage.admin.read().to_field()
+    }
+
+    #[aztec(public)]
+    #[aztec(view)]
+    fn is_minter(minter: AztecAddress) -> bool {
+        storage.minters.at(minter).read()
+    }
+
+    // Returns the total number of NFTs a given address owns in public.
+    #[aztec(public)]
+    #[aztec(view)]
+    fn balance_of_public(owner: AztecAddress) -> Field {
+        storage.public_balances.at(owner).read().to_integer()
+    }
+
+    #[aztec(public)]
+    fn transfer_public(from: AztecAddress, to: AztecAddress, token_id: Field, nonce: Field) {
+        if (!from.eq(context.msg_sender())) {
+            assert_current_call_valid_authwit_public(&mut context, from);
+        } else {
+            assert(nonce == 0, "invalid nonce");
+        }
+
+        ...
+    }
+
+    #[aztec(public)]
+    fn shield(from: AztecAddress, token_id: Field, secret_hash: Field, nonce: Field) {
+        if (!from.eq(context.msg_sender())) {
+            // The redeem is only spendable once, so we need to ensure that you cannot insert multiple shields from the same message.
+            assert_current_call_valid_authwit_public(&mut context, from);
+        } else {
+            assert(nonce == 0, "invalid nonce");
+        }
+
+        ...
+    }
+
+    #[aztec(private)]
+    fn transfer(to: AztecAddress, token_id: Field) {
+        ...
+    }
+
+    #[aztec(private)]
+    fn transfer_from(from: AztecAddress, to: AztecAddress, token_id: Field, nonce: Field) {
+        ...
+    }
+
+    // Burns public NFT
+    #[aztec(public)]
+    fn burn(from: AztecAddress, amount: Field, nonce: Field) {
+        // docs:start:assert_current_call_valid_authwit_public
+        if (!from.eq(context.msg_sender())) {
+            assert_current_call_valid_authwit_public(&mut context, from);
+        } else {
+            assert(nonce == 0, "invalid nonce");
+        }
+        ...
+    }
+
+    #[aztec(private)]
+    fn redeem_shield(to: AztecAddress, amount: Field, secret: Field) {
+        ...
+    }
+
+    #[aztec(private)]
+    fn unshield(from: AztecAddress, to: AztecAddress, amount: Field, nonce: Field) {
+        if (!from.eq(context.msg_sender())) {
+            assert_current_call_valid_authwit(&mut context, from);
+        } else {
+            assert(nonce == 0, "invalid nonce");
+        }
+        ...
+    }
+
+    // Returns zero address when the token does not have public owner. Reverts if the token does not exist.
+    #[aztec(public)]
+    #[aztec(view)]
+    fn owner_of(token_id: Field) -> AztecAddress {
+        storage.public_owners.at(token_id).read()
+    }
```

Comment:
refloating an old slack convo but how do you think offchain proving you have a private NFT would work here?
Mike mentioned once that perhaps we may nee da custom method just to do this?
https://aztecprotocol.slack.com/archives/C04PUD9AA4W/p1719413954061489